runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,728 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   bool isInstanceField = node->getSymbolReference()->getSymbol()->getKind() != TR::Symbol::IsStatic;
+   J9Method *owningMethod = (J9Method *)TR::TreeEvaluator::getMethodFromBCInfo(node->getByteCodeInfo(), cg->comp());
+
+   TR_RuntimeHelper helperIndex = isWrite ? (isInstanceField ? TR_jitReportInstanceFieldWrite: TR_jitReportStaticFieldWrite):
+                                            (isInstanceField ? TR_jitReportInstanceFieldRead: TR_jitReportStaticFieldRead);
+
+   TR::S390HelperLinkage *helperLink = static_cast<TR::S390HelperLinkage*>(cg->getLinkage(runtimeHelperLinkage(helperIndex)));
+   TR::Register *objectReg = isInstanceField ? cg->evaluate(node->getFirstChild()): NULL;
+
+   TR::Register *valueReferenceReg = NULL;
+
+   // In the event that we have to write to a value, the VM helper routine expects that a pointer to the value being written to
+   // is passed in as a parameter. So we must store valueReg into memory and then load the address back into a register in order
+   // to pass the address of that value as an argument. We prepare the register below.
+   if (isWrite)
+      {
+      TR::Node *valueNode = node->getFirstChild();
+      if (isInstanceField)
+         {
+         // Pass in valueNode so it can be set to the correct node.
+         TR::TreeEvaluator::getIndirectWrtbarValueNode(cg, node, valueNode, false);
+         }
+      TR::Register *valueReg = cg->evaluate(valueNode);
+
+      TR::DataType nodeType = valueNode->getDataType();
+      TR::SymbolReference *sr = cg->allocateLocalTemp(nodeType);
+      TR::MemoryReference *valueMR = generateS390MemoryReference(valueNode, sr, cg);
+      if (valueReg->getKind() == TR_GPR)
+         {
+         auto mnemonic = TR::InstOpCode::ST;
+         if (node->getDataType() == TR::Address && TR::Compiler->target.is64Bit() && !cg->comp()->useCompressedPointers())
+            mnemonic = TR::InstOpCode::STG;
+         generateRXInstruction(cg, mnemonic, node, valueReg, valueMR);
+         }
+      else if (valueReg->getKind() == TR_FPR)
+         {
+         auto mnemonic = nodeType == TR::Float ? TR::InstOpCode::STE : TR::InstOpCode::STD;
+         generateRXInstruction(cg, mnemonic, node, valueReg, valueMR);
+         }
+      else
+         {
+         TR_ASSERT_FATAL(false, "Unsupported register kind (%d) for fieldwatch", valueReg->getKind());
+         }
+      // valueReg and valueReferenceReg are different. Add conditions for valueReg here
+      dependencies->addPostCondition(valueReg, TR::RealRegister::AssignAny);
+      valueReferenceReg = cg->allocateRegister();
+      // Now store the stack location into the valueReferenceReg register which will
+      // also be an argument register for the VM helper call.
+      TR::MemoryReference *tempMR = generateS390MemoryReference(*valueMR, 0, cg);
+      generateRXInstruction(cg, TR::InstOpCode::LAY, node, valueReferenceReg , tempMR);
```

Comment:
We probably shouldn't use `LAY` here. We can use `LA`. The API will upgrade to long displacement for us if needed (lazily). Also it seems we don't need to allocate another register for this. Are we simply able to reuse `valueReg`?

Reply:
46037bd

Reply:
This routine was cleaned up in general here: 8e2e97d. `valueReg` should no longer be a concern.